lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp

```diff
@@ -995,6 +1016,41 @@ void LifetimeChecker::handleStoreUse(unsigned UseID) {
     }
   }
 
+  // Check if we're in a cross-module struct initializer that uses RootSelf rather
+  // than DelegatingSelf for Swift 4 compatibility. We look for a problem case by
+  // seeing if there are any assignments to individual fields that might be
+  // initializations; that is, that they're not dominated by `self = other`.
+  auto isFullValueAssignment = [this](const SILInstruction *inst) -> bool {
+    SILType ty;
+    if (auto *assign = dyn_cast<AssignInst>(inst))
+      ty = assign->getSrc()->getType();
+    else if (auto *copyAddr = dyn_cast<CopyAddrInst>(inst))
+      ty = copyAddr->getSrc()->getType().getObjectType();
+    return ty && ty == TheMemory.MemorySILType;
```

Comment:
You should check if the value being assigned to is TheMemory.MemoryInst, instead of just comparing types

Reply:
Oh, of course. I didn't know how to do this and my brain came up with this silly approximation.